### ISSUE TITLE
Get lr scheduler warning when using amp to train with fp16

### ISSUE THREAD

**Issue by CoinCheung** [pytorch/pytorch#27595]
## 🐛 Bug

Have a warning message when I used amp to train a model with fp16.

## To Reproduce

With this code: 
```python
import os
import argparse

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.distributed as dist
from torch.optim.lr_scheduler import _LRScheduler
from apex import amp

from torch.nn import BatchNorm2d




class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv = nn.Conv2d(3, 16, 3, 1, 1)
        self.bn = BatchNorm2d(16)
        self.act = nn.ReLU(inplace=True)
        self.linear = nn.Linear(16, 1000)

    def forward(self, x):
        feat = self.act(self.bn(self.conv(x)))
        feat = torch.mean(feat, dim=(2, 3))
        logits = self.linear(feat)
        return logits


def main():
    model = Model()
    criteria = nn.CrossEntropyLoss()
    model.cuda()

    optimizer = torch.optim.SGD(
        model.parameters(),
        lr=0.016,
        weight_decay=1e-5,
        momentum=0.9
    )
    lr_scheduler = torch.optim.lr_scheduler.StepLR(
        optimizer,
        step_size=10,
    )

    model, optimizer = amp.initialize(model, optimizer, opt_level='O1')

    ims = torch.randn(1, 3, 224, 224).cuda()
    lbs = torch.randint(0, 1000, (1, )).cuda()
    logits = model(ims)
    loss = criteria(logits, lbs)
    loss.backward()
    optimizer.step()
    lr_scheduler.step()


if __name__ == '__main__':
    main()


```

I got the warnining: 

>
/miniconda/envs/py36/lib/python3.6/site-packages/torch/optim/lr_scheduler.py:113: UserWarning: Seems like `optimizer.step()` has been overridden after learning rate scheduler initialization. Please, make sure to call `optimizer.step()` before `lr_scheduler.step()`. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)

If I do not use `amp.initialize`, there will not be this warning message.

cc @vincentqb

**Comment by zou3519** [comment]
Unsure if amp bug or PyTorch bug.

**Comment by gchanan** [comment]
CC @mcarilli.

**Comment by gchanan** [comment]
marking dependency bug until we hear otherwise.

**Comment by mcarilli** [comment]
@CoinCheung Amp patches the `step` method to skip the step whenever inf/nan gradients are encountered.  Therefore, with `opt_level=O1`, I expect you should see this warning intermittently (for the occasional iterations where an inf/nan gradient is encountered, including the first few iterations as the scale calibrates).  I don't think it actually breaks anything.

**Comment by SebastienEske** [comment]
Hello, I have a similar issue which hapenning without amp. The way I see it, it is due to the scheduler initialization.

`scheduler.step()` is supposed to be used after `optimizer.step()` as indicated [here](https://github.com/pytorch/pytorch/blob/19d83ab8007aa5f295b3d69e8283bfbb994a011a/torch/optim/lr_scheduler.py#L45).

But [a few lines below](https://github.com/pytorch/pytorch/blob/19d83ab8007aa5f295b3d69e8283bfbb994a011a/torch/optim/lr_scheduler.py#L77) `scheduler.step()` is called without any call to the optimizer.

If think it should not be called. What do you think? Am I misunderstanding it?

**Comment by oersoy1** [comment]
Hi all, @mcarilli  upping this thread, if you don't mind. I was trying to figure out where this warning was coming from. I am using huggingfaces's transformers for some NLP tasjs and I thought it was there since I made a version update. Then I thought it might be Pytorch, but it turns out it is apex.  Apex's F16 mixed-precision gives me a compute power boost obviously without much accuracy loss. Has anyone checked with NVIDIA guys to make sure it is not breaking anything?  In my logs, which is part of my deliverables, it clears says I am an idiot, I don't know how to sequence my scheduler/optimizer.... Lol...

**Comment by timocato** [comment]
Simple, do the following and this warning will gone:  

change your code from: 

    optimizer = torch.optim.SGD(
        model.parameters(),
        lr=0.016,
        weight_decay=1e-5,
        momentum=0.9
    )
    lr_scheduler = torch.optim.lr_scheduler.StepLR(
        optimizer,
        step_size=10,
    )

    model, optimizer = amp.initialize(model, optimizer, opt_level='O1')


to this: 


    optimizer = torch.optim.SGD(
        model.parameters(),
        lr=0.016,
        weight_decay=1e-5,
        momentum=0.9
    )

    model, optimizer = amp.initialize(model, optimizer, opt_level='O1')

    lr_scheduler = torch.optim.lr_scheduler.StepLR(
        optimizer,
        step_size=10,
    )

**Comment by CoinCheung** [comment]
@timocato Since pytorch supports mixed precision natively now. Do you know how to solve this problem with pytorch native mixed precision training?
